### PR TITLE
Naji/proxy peering client

### PR DESCRIPTION
**Pull request by NajiObeid**
https://github.com/gravitational/cloud/issues/1306

### CODE REVIEW

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:52]: I am thinking we will need to get the RootCAs here similar to what we are doing on the server side to handle certificate rotation. Based on [this](https://github.com/golang/go/issues/22836) issue it seems like there is no `GetConfigForClient` equivalent on the client side of things. But there are some other ideas to accomplish the same thing in the comments of that issue.

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:228]: Creating a new connection while we hold the lock could cause some problems.

Also, I think ideally we should try to set up connections before hand rather than as needed. This would avoid slow session starts for the first connection between proxies. But maybe this is a premature optimization? 

The idea would be to use the Cache/Watcher to be able to create/cleanup connections as proxies are added and deleted. This could also help to avoid removing a client that happens to be in a temporary `TransientFailure` state during cleanup.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:52]: Yes I briefly looked into it, that's why I had the access cache added to the config. I'll dig in deeper and see what I can do

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:228]: > Also, I think ideally we should try to set up connections before hand rather than as needed. This would avoid slow session starts for the first connection between proxies. But maybe this is a premature optimization?

No it's all good. I thought we weren't going to include the watcher in this library, and have him call this `dial` function after it got exported. But it you know what now that you mention it, it will look cleaner. I'll work on it.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:52]: 2df5b1670585c716e35e18fad703f6e12c0cee41

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:235]: I don't think we need to test the connection. The grpc.ClientConn should reconnect automatically if there is a transient failure. However we may want to check if the PeerAddr has changed and close/reconnect in that case.

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:250]: Could we avoid holding the global lock while testing/creating the connection? This would prevent calls to DialNode from getting through while we setup connections to other proxies.

Comment (dboslee) [inline lib/proxy/client.go:250]: Could we avoid holding the global lock while testing/creating the connection? This would prevent calls to DialNode from getting through while we setup connections to other proxies.

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:235]: I don't think we need to test the connection. The `grpc.ClientConn` should reconnect automatically if there is a transient failure. However we may want to check if the `PeerAddr` has changed and close/reconnect in that case.

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:268]: Do we need this extra cleanup step? The connection cleanup seems to be handled in `updateConnections`. Also as I mentioned above event if `testConnection` fails the `grpc.ClientConn` will handle reconnecting a transient failure.

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:353]: `closeConn` is called from a few different places above while holding the global lock could this block for a while? Maybe we can defer graceful shutdowns to be handled by the reverse tunnel server or where ever this client library is used.

**Review by dboslee — COMMENTED** [review]
Mostly concerned with locking while making network requests like connecting/testing/closing the connection. I think `grpc.DialContext` is non blocking unless you pass in a specific `DialOption` and it seems `clientConn.Close` is non blocking as well based on [this discussion](https://github.com/grpc/grpc-go/issues/1726). So if we drop testing the connection I think we should be all set.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:268]: It takes a long time for the watcher to notice a proxy is no longer available. But it does a good job of noticing newly added proxies. At least that's what I saw during testing. 
So the only real benefit for it is metrics, as the alternative would be to fail when Dial is called.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:353]: I've only tested this on a small scale and while writing tests, so don't have a good answer about performance or holding the lock for an extended period of time, but I am going to investigate and try to get some numbers or stress test that somehow.
But the issue mainly comes from the buffering grpc does in the background. I just wanted to be able to test cases such as:
- closed client connections -> streams should error out and not accept any more send messages
- streams didn't always receive eof errors a connection is dropped, so we couldn't log this properly in prometheus.
...

that's mainly why I elected to closestreams manually before finalizing closing the connection. I am not sure it does affect things in real life, but I found lots of complaints about that on the grpc-go git repo, mostly people asking for graceful shutdown on the client side just like there is one on the server side

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:235]: Oh ok, I assumed that for the proxy peer addr to change the proxy needs to be restarted, and assumed it gets a new id after a restart so it wouldn't be necessary to check for that case. 

as for the connection testing, the connection state isn't always up to date and we need to do something with it for GetState to return the correct one, that's why testConnection is there. Maybe I should check the error returned or the state right after and keep the connection in case it's a transient failure and not something more permanent.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:250]: I'll try to optimize it a bit, but it depends what's the experience we want to present in case of concurrent calls to Dial and update... fail because the cache isn't up to date or wait...

Comment (dboslee) [reply lib/proxy/client.go:250]: I think failing if the cache is out of date is fine.

File: lib/proxy/client.go
Comment (dboslee) [reply lib/proxy/client.go:235]: For cloud the peer address changing probably isn't an issue but could be in other environments.

File: lib/proxy/client.go
Comment (dboslee) [reply lib/proxy/client.go:268]: I think this is due to the the ttl set in the cache/backend. Once a proxy is disconnected it stays around in the cache/backend for something like 10-20 minutes.

What is the problem, that the `tunnelErrorCounter` would be lower until the cache is updated and the client gets closed? 

We can't really be sure whether its a transient failure or the proxy was permanently disconnected until some arbitrary timeout anyways so why not use the cache/backend as the source of truth for this?

And this relates to your comment above...

> as for the connection testing, the connection state isn't always up to date and we need to do something with it for GetState to return the correct one, that's why testConnection is there. Maybe I should check the error returned or the state right after and keep the connection in case it's a transient failure and not something more permanent.

If we wait for the proxy to expire in the backend/cache we don't need to worry about checking the state. And when we are in this state where the proxy is permanently disconnected but still in the cache, I think the failure would happen when establishing/sending on the stream. Maybe it would show as a timeout

**Comment by NajiObeid** [comment]
> I think grpc.DialContext is non blocking unless you pass in a specific DialOption and it seems clientConn.Close is non blocking as well based on https://github.com/grpc/grpc-go/issues/1726. So if we drop testing the connection I think we should be all set.

The fact that both are asynchronous makes it impossible for us to write proper tests. Also dialling a connection and testing it right after fails much faster than adding the WithBlock() Dial option to the connection setting in combination with an expirable context.

**Comment by russjones** [comment]
@NajiObeid  Is this PR ready for review?

**Comment by NajiObeid** [comment]
@russjones yes it is! thanks

File: lib/proxy/clientinterceptor.go
Comment (rosstimothy) [inline lib/proxy/clientinterceptor.go:26]: Did you consider using a `stats.Handler`? 

https://pkg.go.dev/google.golang.org/grpc/stats#Handler

File: lib/proxy/auth.go
Comment (NajiObeid) [reply lib/proxy/auth.go:104]: According to the docs

`// If the returned Config is nil, the original Config will be used.`

I've seen it implemented like this other parts of the codebase so I kept it. I can change it to be more consistent with go standards

File: lib/proxy/clientinterceptor.go
Comment (NajiObeid) [reply lib/proxy/clientinterceptor.go:31]: I didn't know this existed :smile: But I would still use a wrapper around it to check for io.EOF

File: lib/proxy/clientinterceptor.go
Comment (NajiObeid) [reply lib/proxy/clientinterceptor.go:26]: The interceptor was initially added to capture and convert errors to grpc type errors. I then thought to add the metrics in the interceptor itself since the infrastructure is already there. But you're right stats.Handler is looks cleaner and is easier to follow. If you have any strong opinions on it, I would move the metrics out of the interceptor, it shouldn't be too much work

Comment (rosstimothy) [reply lib/proxy/clientinterceptor.go:26]: I don't necessarily have a strong opinion about it. While looking at some of the code added to handle metrics I wondered if some of it could be eliminated/simplified by using a `stats.Handler`. Most implementations tracking grpc metrics that I have seen have used a `stats.Handler`, cockroachdb for example: https://github.com/cockroachdb/cockroach/blob/4d987b3f0788c98fa358715b336a1c55e254ee50/pkg/rpc/stats_handler.go

File: lib/proxy/clientinterceptor.go
Comment (rosstimothy) [reply lib/proxy/clientinterceptor.go:31]: Perhaps we should upstream the `io.EOF` check to `trail.ToGRPC`?

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:436]: The issue I found with blocking and context timeouts is that it takes until the context expires to report failed dials, while the alternative of asynchronously connecting and testing the connection right after fails much faster

**Comment by NajiObeid** [comment]
I'll also bump the trace library version once that's also merged

File: lib/utils/grpc.go
Comment (fspmarshall) [inline lib/utils/grpc.go:42]: I'm under the (potentially wrong) impression that the errors returned by `SendMsg` and `RecvMsg` are errors being bubbled up from the GRPC library to the calling code (i.e. teleport), in which case you'd want to be using `trail.FromGRPC` if anything.  I got that impression based on poking around in some generated code that I don't fully understand, so take it with a grain of salt.

Either way, this definitely deserves some test coverage to verify that the `trace` error types are being preserved correctly across the wire.

File: lib/proxy/helpers_test.go
Comment (rosstimothy) [inline lib/proxy/helpers_test.go:202]:
```suggestion
	t.Cleanup(func({ require.NoError(t, server.Close()) })
```

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:246]: [8bf7a15](https://github.com/gravitational/teleport/pull/10440/commits/8bf7a15417c7d764db93e7eeb2252f8dae0abd8a)

File: lib/proxy/clientmetrics.go
Comment (rosstimothy) [inline lib/proxy/clientmetrics.go:112]:
```suggestion
```

File: go.mod
Comment (rosstimothy) [inline go.mod:264]: I don't see these being used anywhere. Was adding them on purpose?

File: lib/proxy/clientmetrics.go
Comment (rosstimothy) [inline lib/proxy/clientmetrics.go:44]:
```suggestion
				Help:      "Total number of errors encountered dialing peer proxies.",
```

File: lib/proxy/client.go
Comment (rosstimothy) [inline lib/proxy/client.go:525]: Should this also check the connectivity state and report an `errorTransientFailure` like in `testConnection`? Or is that error only meant to be recorded during the connection tests? If it's not just meant for `testConnection` there might be more places than this which should be checking the connectivity state

File: go.mod
Comment (NajiObeid) [reply go.mod]: not sure why it got messed up during the rebase. cleaned it now

File: lib/proxy/client.go
Comment (rosstimothy) [reply lib/proxy/client.go:525]: I think the logic is fine - I was more just questioning if the tunnel error being reported should be more reflective here. 
```suggestion
	      if conn.GetState() == connectivity.TransientFailure {
			c.metrics.reportTunnelError(errorTransientFailure)
		} else {
		        c.metrics.reportTunnelError(errorStreamStart)
		}
```

I guess what I'm trying to ask is would being more explicit here about what gets reported aid in debugging/diagnosing issues? Although I suppose if there is an error returned by `DialNode` there is a pretty good chance the connectivity state is `TransientFailure`

Also now that I think about it more should `errorTransientFailure` be something more descriptive and less grpc-like. Maybe something like `errorConnectionFailure = "PEER_CONNECTION_FAILURE` ?

File: lib/proxy/clientmetrics.go
Comment (dboslee) [inline lib/proxy/clientmetrics.go:89]: This goes for both the client and server side. Why did you choose a histogram over a counter for `messageSent` and `messageReceived`?

File: lib/proxy/clientmetrics.go
Comment (dboslee) [reply lib/proxy/clientmetrics.go:49]: Probably should track the `idle` state as well. But the `connecting` state can be lumped together with `transient_failure` in my opinion.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:492]: Yes I remember you mentioned this earlier. I haven't got the chance to test finish testing this properly yet. Thanks for the reminder.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:541]: You're right it is not necessary, but it's mostly useful for reporting:
when dialling a new proxy, the only ways for us to know and report connection issues are:
- add the `WaitForReady` flag in the dial options. That waits for the duration set for timeouts before reporting unreachable servers, and that can be too long when dialing a proxy through updates or individually triggered when dialing a proxy that's not in the cache.
- without the `WaitForReady` flag. We'd have to wait for the first time a connection is being used to report it hasn't been established properly or have a routine monitoring connections.

Well I chose to initiate that test manually to get a sense if the connection has been established or not 😄. We can go with the monitoring routine if you feel that's better or even do the testing in a separate routine just for reporting purposes...

File: lib/proxy/clientmetrics.go
Comment (NajiObeid) [reply lib/proxy/clientmetrics.go:49]: let me see if I have access to that kind of info in the stats handler where this is reported

File: lib/proxy/clientmetrics.go
Comment (NajiObeid) [reply lib/proxy/clientmetrics.go:89]: Ah the variable name relays the wrong info. We're not counting the number of messages sent and received, I don't think we care about that. We're just reporting the size. Unless you meant why we're not reporting the size in a counter in your question?

File: lib/proxy/client.go
Comment (dboslee) [reply lib/proxy/client.go:541]: I think the goroutine to monitor things would be better. It could just loop at some interval checking the state of each connection.

The main benefit I see is that if we discover several new proxies at once they can be added to the map faster rather than waiting while each connection is established and tested.

Also we would probably want a goroutine to monitor them regardless of whether we test them or not. Otherwise we would only update the metrics when a new set of proxies comes in or we attempt to dial over the connection.

File: lib/proxy/clientmetrics.go
Comment (dboslee) [reply lib/proxy/clientmetrics.go:89]: > Unless you meant why we're not reporting the size in a counter in your question?

This is what I meant. Whats the benefits of a histogram vs counter here?

Comment (NajiObeid) [reply lib/proxy/clientmetrics.go:89]: I can see this working ok with a counter, but isn't this kind of data more suitable for a histogram?

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:541]: > Also we would probably want a goroutine to monitor them regardless of whether we test them or not. Otherwise we would only update the metrics when a new set of proxies comes in or we attempt to dial over the connection.

Now that I think about this more, I kind of disagree with it a bit. 
- It is possible we miss reporting connections being down if the ticker delay isn't small enough. In that case what would you think would be an acceptable interval?
- We're also probably not interested in these kinds of metrics, at least I am not 😄. What I care most about is mostly of the following order; were we able to dial a proxy we've been told is reachable? 

But on the other side I don't have a really strong opinion on either approaches. I'll rework the code a bit and see how it plays with the `loadBalancingPolicy` config

File: lib/proxy/clientmetrics.go
Comment (dboslee) [reply lib/proxy/clientmetrics.go:89]: > the total sum of all observed values, exposed as <basename>_sum
Ah I didn't know that histograms also come with a sum. So that works for me

**Comment by pierrebeaucamp** [comment]
@rosstimothy / @fspmarshall / @espadolini can we get another review please?

File: lib/proxy/client_test.go
Comment (rosstimothy) [inline lib/proxy/client_test.go:1]: It looks like all the tests here are exercising private functions. It would be nice to see a few integration tests that have Proxies/Nodes and actually call `Dial`

Comment (NajiObeid) [reply lib/proxy/client_test.go:1]: I have this template that tests the old behaviour here https://github.com/gravitational/teleport/pull/11671
Once all the changes are in it should be as simple as running a pretty similar test with the new config.

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:314]: Sorry I grew up with british spelling, not that I am good at it but it's hard to let go of old habits 😄 I think you've mentioned that a couple of times already

File: go.mod
Comment (NajiObeid) [reply go.mod:267]: I'll clean that

File: lib/proxy/client.go
Comment (rosstimothy) [reply lib/proxy/client.go:524]: ah sorry I didn't follow that far enough down to get to the version

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go]: [95f361a4ffcc966c696f25ab8675d519e4667bf0](https://github.com/gravitational/teleport/pull/9132/commits/95f361a4ffcc966c696f25ab8675d519e4667bf0)

File: lib/proxy/client.go
Comment (rosstimothy) [inline lib/proxy/client.go:305]: nit: forgot to update godoc
```suggestion
// DialNode dials a node through a peer proxy.
```

**Review by espadolini — COMMENTED** [review]
Spurious comments that I got while reviewing #10444 (not a full pass).

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:488]: Good catch .we can improve that a bit by delaying the `append` after the connections have been made.

File: lib/proxy/client.go
Comment (rosstimothy) [inline lib/proxy/client.go:73]: Is there a reason for making this private? 99% of these are public in the codebase and it seems a bit off to me that `ClientConfig` is public yet I wouldn't be able to verify it was populated properly by calling this.

File: lib/proxy/client.go
Comment (rosstimothy) [inline lib/proxy/client.go:241]: nit: This is 40 lines north of where it is first used, can it be moved a little closer?

File: lib/proxy/client.go
Comment (NajiObeid) [reply lib/proxy/client.go:73]: There's no reason at all, I must have written it without checking how it's done elsewhere.
But it is my understanding that calling `CheckAndSetDefaults` is implied in every `NewSomething` call. `ClientConfig` is worthless on its own without the `NewClient` call, so I wouldn't expect anyone to make the check manually. Unless services start sharing the same config structs or the function is not automatically called when initializing a new object, I don't see why it should be public. Am I missing something?

**Review by fspmarshall — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Review by dboslee — APPROVED** [review]

**Pull request merged**
